File: windows/QMK Toolbox/Flashing.cs
using QMK_Toolbox.Helpers;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Windows.Forms;

namespace QMK_Toolbox
{
    public enum Chipset
    {
        Apm32Dfu,
        AtmelDfu,
        AtmelSamBa,
        AvrIsp,
        BootloadHid,
        Caterina,
        Halfkay,
        Kiibohd,
        LufaMs,
        QmkDfu,
        Stm32Dfu,
        Stm32Duino,
        UsbAsp,
        UsbTiny,
        NumberOfChipsets
    };

    public class Flashing : EventArgs
    {
        private readonly Process _process;
        private readonly ProcessStartInfo _startInfo;

        public string ComPort = "";
        public string MountPoint = "";

        private readonly Printing _printer;
        public Usb Usb;

        private readonly string[] _resources = {
            "avrdude.conf",
            "reset.eep",
            "reset_left.eep",
            "reset_right.eep",
            "avrdude.exe",
            "bootloadHID.exe",
            "dfu-programmer.exe",
            "dfu-util.exe",
            "mdloader.exe",
            "teensy_loader_cli.exe",
            "libftdi1.dll",
            "libusb0.dll",
            "libusb-0-1-4.dll",
            "libusb-1.0.dll",
            "libwinpthread-1.dll"
        };

        public Flashing(Printing printer)
        {
            _printer = printer;
            EmbeddedResourceHelper.ExtractResources(_resources);

            _process = new Process();
            _startInfo = new ProcessStartInfo
            {
                UseShellExecute = false,
                RedirectStandardError = true,
                RedirectStandardOutput = true,
                RedirectStandardInput = true,
                CreateNoWindow = true
            };
        }

        private void OnOutputDataReceived(object sender, DataReceivedEventArgs e)
        {
            Debug.Write(e.Data);
            _printer.PrintResponse(e.Data, MessageType.Info);
        }

        private void OnErrorDataReceived(object sender, DataReceivedEventArgs e)
        {
            Debug.Write(e.Data);
            _printer.PrintResponse(e.Data, MessageType.Info);
        }

        private void ProcessOutput(object streamReader)
        {
            StreamReader _stream = (StreamReader)streamReader;
            string output;

            while (!_stream.EndOfStream)
            {
                output = _stream.ReadLine() + "\n";
                _printer.PrintResponse(output, MessageType.Info);

                if (output.Contains("Bootloader and code overlap.") || // DFU
                    output.Contains("exceeds remaining flash size!") || // BootloadHID
                    output.Contains("Not enough bytes in device info report")) // BootloadHID
                {
                    _printer.Print("File is too large for device", MessageType.Error);
                }
            }
        }
        private void RunProcess(string command, string args)
        {
            _printer.Print($"{command} {args}", MessageType.Command);
            _startInfo.WorkingDirectory = Application.LocalUserAppDataPath;
            _startInfo.FileName = Path.Combine(Application.LocalUserAppDataPath, command);
            _startInfo.Arguments = args;
            _startInfo.RedirectStandardOutput = true;
            _startInfo.RedirectStandardError = true;
            _startInfo.UseShellExecute = false;
            _process.StartInfo = _startInfo;

            _process.Start();

            // Thread that handles STDOUT
            Thread _ThreadProcessOutput = new Thread(ProcessOutput);
            _ThreadProcessOutput.Start(_process.StandardOutput);

            // Thread that handles STDERR
            _ThreadProcessOutput = new Thread(ProcessOutput);
            _ThreadProcessOutput.Start(_process.StandardError);

            _process.WaitForExit();
        }

        public void Flash(string mcu, string file)
        {
            if (Usb.CanFlash(Chipset.AtmelDfu) || Usb.CanFlash(Chipset.QmkDfu))
                FlashAtmelDfu(mcu, file);
            if (Usb.CanFlash(Chipset.Caterina))
                FlashCaterina(mcu, file);
            if (Usb.CanFlash(Chipset.Halfkay))
                FlashHalfkay(mcu, file);
            if (Usb.CanFlash(Chipset.Stm32Dfu))
                FlashStm32Dfu(file);
            if (Usb.CanFlash(Chipset.Apm32Dfu))
                FlashApm32Dfu(file);
            if (Usb.CanFlash(Chipset.Kiibohd))
                FlashKiibohd(file);
            if (Usb.CanFlash(Chipset.LufaMs))
                FlashLufaMs(file);
            if (Usb.CanFlash(Chipset.AvrIsp))
                FlashAvrIsp(mcu, file);
            if (Usb.CanFlash(Chipset.UsbAsp))
                FlashUsbAsp(mcu, file);
            if (Usb.CanFlash(Chipset.UsbTiny))
                FlashUsbTiny(mcu, file);
            if (Usb.CanFlash(Chipset.BootloadHid))
                FlashBootloadHid(file);
            if (Usb.CanFlash(Chipset.AtmelSamBa))
                FlashAtmelSamBa(file);
            if (Usb.CanFlash(Chipset.Stm32Duino))
                FlashStm32Duino(file);
        }

        public void Reset(string mcu)
        {
            if (Usb.CanFlash(Chipset.AtmelDfu) || Usb.CanFlash(Chipset.QmkDfu))
                ResetAtmelDfu(mcu);
            if (Usb.CanFlash(Chipset.Halfkay))
                ResetHalfkay(mcu);
            if (Usb.CanFlash(Chipset.BootloadHid))
                ResetBootloadHid();
            if (Usb.CanFlash(Chipset.AtmelSamBa))
                ResetAtmelSamBa();
        }

        public void ClearEeprom(string mcu)
        {
            if (Usb.CanFlash(Chipset.AtmelDfu) || Usb.CanFlash(Chipset.QmkDfu))
                ClearEepromAtmelDfu(mcu, !Usb.CanFlash(Chipset.QmkDfu));
            if (Usb.CanFlash(Chipset.Caterina))
                ClearEepromCaterina(mcu);
            if (Usb.CanFlash(Chipset.UsbAsp))
                ClearEepromUsbAsp(mcu);
        }

        public void SetHandedness(string mcu, bool rightHand)
        {
            if (Usb.CanFlash(Chipset.AtmelDfu) || Usb.CanFlash(Chipset.QmkDfu))
                SetHandednessAtmelDfu(mcu, rightHand, !Usb.CanFlash(Chipset.QmkDfu));
            if (Usb.CanFlash(Chipset.Caterina))
                SetHandednessCaterina(mcu, rightHand);
            if (Usb.CanFlash(Chipset.UsbAsp))
                SetHandednessUsbAsp(mcu, rightHand);
        }

        public bool CanFlash() => Usb.AreDevicesAvailable();

        public bool CanReset()
        {
            var resettable = new List<Chipset> {
                Chipset.AtmelDfu,
                Chipset.AtmelSamBa,
                Chipset.BootloadHid,
                Chipset.Halfkay,
                Chipset.QmkDfu
            };
            foreach (Chipset chipset in resettable)
            {
                if (Usb.CanFlash(chipset))
                    return true;
            }
            return false;
        }

        public bool CanClearEeprom()
        {
            var clearable = new List<Chipset>
            {
                Chipset.AtmelDfu,
                Chipset.Caterina,
                Chipset.QmkDfu,
                Chipset.UsbAsp
            };
            foreach (Chipset chipset in clearable)
            {
                if (Usb.CanFlash(chipset))
                    return true;
            }
            return false;
        }

        private void FlashApm32Dfu(string file)
        {
            if (Path.GetExtension(file)?.ToLower() == ".bin")
            {
                RunProcess("dfu-util.exe", $"-a 0 -d 314B:0106 -s 0x08000000:leave -D \"{file}\"");
            }
            else
            {
                _printer.Print("Only firmware files in .bin format can be flashed with dfu-util!", MessageType.Error);
            }
        }

        private void FlashAtmelDfu(string mcu, string file)
        {
            RunProcess("dfu-programmer.exe", $"{mcu} erase --force");
            RunProcess("dfu-programmer.exe", $"{mcu} flash --force \"{file}\"");
            RunProcess("dfu-programmer.exe", $"{mcu} reset");
        }

        private void ResetAtmelDfu(string mcu) => RunProcess("dfu-programmer.exe", $"{mcu} reset");

        private void ClearEepromAtmelDfu(string mcu, bool erase)
        {
            if (erase)
            {
                RunProcess("dfu-programmer.exe", $"{mcu} erase --force");
            }
            RunProcess("dfu-programmer.exe", $"{mcu} flash --force --suppress-validation --eeprom reset.eep");
            if (erase)
            {
                _printer.Print("Please reflash device with firmware now", MessageType.Bootloader);
            }
        }

        private void SetHandednessAtmelDfu(string mcu, bool rightHand, bool erase)
        {
            if (erase)
            {
                RunProcess("dfu-programmer.exe", $"{mcu} erase --force");
            }
            RunProcess("dfu-programmer.exe", $"{mcu} flash --force --suppress-validation --eeprom reset_{(rightHand ? "right" : "left")}.eep");
            if (erase)
            {
                _printer.Print("Please reflash device with firmware now", MessageType.Bootloader);
            }
        }

        private void FlashAtmelSamBa(string file) => RunProcess("mdloader.exe", $"-p {ComPort} -D \"{file}\" --restart");

        private void ResetAtmelSamBa() => RunProcess("mdloader.exe", $"-p {ComPort} --restart");

        private void FlashAvrIsp(string mcu, string file)
        {
            RunProcess("avrdude.exe", $"-p {mcu} -c avrisp -U flash:w:\"{file}\":i -P {ComPort}");
            _printer.Print("Flash complete", MessageType.Bootloader);
        }

        private void FlashBootloadHid(string file) => RunProcess("bootloadHID.exe", $"-r \"{file}\"");

        private void ResetBootloadHid() => RunProcess("bootloadHID.exe", $"-r");

        private void FlashCaterina(string mcu, string file) => RunProcess("avrdude.exe", $"-p {mcu} -c avr109 -U flash:w:\"{file}\":i -P {ComPort}");

        private void ClearEepromCaterina(string mcu) => RunProcess("avrdude.exe", $"-p {mcu} -c avr109 -U eeprom:w:reset.eep:i -P {ComPort}");

        private void SetHandednessCaterina(string mcu, bool rightHand) => RunProcess("avrdude.exe", $"-p {mcu} -c avr109 -U eeprom:w:reset_{(rightHand ? "right" : "left")}.eep:i -P {ComPort}");

        private void FlashHalfkay(string mcu, string file) => RunProcess("teensy_loader_cli.exe", $"-mmcu={mcu} \"{file}\" -v");

        private void ResetHalfkay(string mcu) => RunProcess("teensy_loader_cli.exe", $"-mmcu={mcu} -bv");

        private void FlashKiibohd(string file)
        {
            if (Path.GetExtension(file)?.ToLower() == ".bin")
            {
                RunProcess("dfu-util.exe", $"-D \"{file}\"");
            }
            else
            {
                _printer.Print("Only firmware files in .bin format can be flashed with dfu-util!", MessageType.Error);
            }
        }

        private void FlashLufaMs(string file)
        {
            if (MountPoint != null)
            {
                if (Path.GetExtension(file)?.ToLower() == ".bin")
                {
                    var destFile = $"{MountPoint}\\FLASH.BIN";

                    try
                    {
                        _printer.Print($"Deleting {destFile}...", MessageType.Command);
                        File.Delete(destFile);

                        _printer.Print($"Copying {file} to {destFile}...", MessageType.Command);
                        File.Copy(file, destFile);

                        _printer.Print("Done, please eject drive now.", MessageType.Info);
                    }
                    catch (IOException e)
                    {
                        _printer.Print($"IO ERROR: {e.Message}", MessageType.Error);
                    }
                }
                else
                {
                    _printer.Print("Only firmware files in .bin format can be flashed with this bootloader!", MessageType.Error);
                }
            }
            else
            {
                _printer.Print("Could not find drive letter for device!", MessageType.Error);
            }
        }

        private void FlashStm32Dfu(string file)
        {
            if (Path.GetExtension(file)?.ToLower() == ".bin")
            {
                RunProcess("dfu-util.exe", $"-a 0 -d 0483:DF11 -s 0x08000000:leave -D \"{file}\"");
            }
            else
            {
                _printer.Print("Only firmware files in .bin format can be flashed with dfu-util!", MessageType.Error);
            }
        }

        private void FlashStm32Duino(string file)
        {
            if (Path.GetExtension(file)?.ToLower() == ".bin")
            {
                RunProcess("dfu-util.exe", $"-a 2 -d 1EAF:0003 -R -D \"{file}\"");
            }
            else
            {
                _printer.Print("Only firmware files in .bin format can be flashed with dfu-util!", MessageType.Error);
            }
        }

        private void FlashUsbAsp(string mcu, string file)
        {
            RunProcess("avrdude.exe", $"-p {mcu} -c usbasp -U flash:w:\"{file}\":i");
            _printer.Print("Flash complete", MessageType.Bootloader);
        }

        private void ClearEepromUsbAsp(string mcu) => RunProcess("avrdude.exe", $"-p {mcu} -c usbasp -U eeprom:w:reset.eep:i");

        private void SetHandednessUsbAsp(string mcu, bool rightHand) => RunProcess("avrdude.exe", $"-p {mcu} -c usbasp -U eeprom:w:reset_{(rightHand ? "right" : "left")}.eep:i");

        private void FlashUsbTiny(string mcu, string file)
        {
            RunProcess("avrdude.exe", $"-p {mcu} -c usbtiny -U flash:w:\"{file}\":i");
            _printer.Print("Flash complete", MessageType.Bootloader);
        }
    }
}

File: macos/QMK Toolbox/Flashing.m
#import "Flashing.h"

#import "USB.h"

@interface Flashing ()
@property Printing *printer;
@end

@implementation Flashing
@synthesize serialPort;
@synthesize mountPoint;

- (id)initWithPrinter:(Printing *)p {
    if (self = [super init]) {
        _printer = p;
    }
    return self;
}

- (NSString *)runProcess:(NSString *)command withArgs:(NSArray<NSString *> *)args {
    [self.printer print:[NSString stringWithFormat:@"%@ %@", command, [args componentsJoinedByString:@" "]] withType:MessageType_Command];
    //int pid = [[NSProcessInfo processInfo] processIdentifier];
    NSPipe *pipe = [NSPipe pipe];
    NSFileHandle *file = pipe.fileHandleForReading;

    NSTask *task = [[NSTask alloc] init];
    task.launchPath = [[NSBundle mainBundle] pathForResource:command ofType:@""];
    task.currentDirectoryPath = [[NSBundle mainBundle] resourcePath];
    task.arguments = args;
    task.standardOutput = pipe;
    task.standardError = pipe;

    [task launch];

    NSData *data = [file readDataToEndOfFile];
    [file closeFile];

    NSString *grepOutput = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    // NSLog (@"grep returned:\n%@", grepOutput);
    [self.printer printResponse:grepOutput withType:MessageType_Command];
    return grepOutput;
}

- (void)flash:(NSString *)mcu withFile:(NSString *)file {
    if ([USB canFlash:AtmelDFU] || [USB canFlash:QMKDFU])
        [self flashAtmelDFU:mcu withFile:file];
    if ([USB canFlash:Caterina])
        [self flashCaterina:mcu withFile:file];
    if ([USB canFlash:Halfkay])
        [self flashHalfkay:mcu withFile:file];
    if ([USB canFlash:STM32DFU])
        [self flashSTM32DFUWithFile:file];
    if ([USB canFlash:APM32DFU])
        [self flashAPM32DFUWithFile:file];
    if ([USB canFlash:Kiibohd])
        [self flashKiibohdWithFile:file];
    if ([USB canFlash:STM32Duino])
        [self flashSTM32DuinoWithFile:file];
    if ([USB canFlash:AVRISP])
        [self flashAVRISP:mcu withFile:file];
    if ([USB canFlash:USBAsp])
        [self flashUSBAsp:mcu withFile:file];
    if ([USB canFlash:USBTiny])
        [self flashUSBTiny:mcu withFile:file];
    if ([USB canFlash:AtmelSAMBA])
        [self flashAtmelSAMBAwithFile:file];
    if ([USB canFlash:BootloadHID])
        [self flashBootloadHIDwithFile:file];
    if ([USB canFlash:LUFAMS])
        [self flashLUFAMSwithFile:file];
}

- (void)reset:(NSString *)mcu {
    if ([USB canFlash:AtmelDFU] || [USB canFlash:QMKDFU])
        [self resetAtmelDFU:mcu];
    if ([USB canFlash:Halfkay])
        [self resetHalfkay:mcu];
    if ([USB canFlash:AtmelSAMBA])
        [self resetAtmelSAMBA];
    if ([USB canFlash:BootloadHID])
        [self resetBootloadHID];
}

- (void)clearEEPROM:(NSString *)mcu {
    if ([USB canFlash:AtmelDFU] || [USB canFlash:QMKDFU])
        [self clearEEPROMAtmelDFU:mcu eraseFirst:![USB canFlash:QMKDFU]];
    if ([USB canFlash:Caterina])
        [self clearEEPROMCaterina:mcu];
    if ([USB canFlash:USBAsp])
        [self clearEEPROMUSBAsp:mcu];
}

- (void)setHandedness:(NSString *)mcu rightHand:(BOOL)rightHand {
    if ([USB canFlash:AtmelDFU] || [USB canFlash:QMKDFU])
        [self setHandednessAtmelDFU:mcu rightHand:rightHand eraseFirst:![USB canFlash:QMKDFU]];
    if ([USB canFlash:Caterina])
        [self setHandednessCaterina:mcu rightHand:rightHand];
    if ([USB canFlash:USBAsp])
        [self setHandednessUSBAsp:mcu rightHand:rightHand];
}

- (BOOL)canFlash {
    return [USB areDevicesAvailable];
}

- (BOOL)canReset {
    NSArray<NSNumber *> *resettable = @[
        @(AtmelDFU),
        @(AtmelSAMBA),
        @(BootloadHID),
        @(Halfkay),
        @(QMKDFU)
    ];
    for (NSNumber *chipset in resettable) {
        if ([USB canFlash:(Chipset)chipset.intValue])
            return YES;
    }
    return NO;
}

- (BOOL)canClearEEPROM {
    NSArray<NSNumber *> *clearable = @[
        @(AtmelDFU),
        @(Caterina),
        @(QMKDFU),
        @(USBAsp)
    ];
    for (NSNumber *chipset in clearable) {
        if ([USB canFlash:(Chipset)chipset.intValue])
            return YES;
    }
    return NO;
}

- (void)flashAPM32DFUWithFile:(NSString *)file {
    if([[[file pathExtension] lowercaseString] isEqualToString:@"bin"]) {
        [self runProcess:@"dfu-util" withArgs:@[@"-a", @"0", @"-d", @"314B:0106", @"-s", @"0x8000000:leave", @"-D", file]];
    } else {
        [self.printer print:@"Only firmware files in .bin format can be flashed with dfu-util!" withType:MessageType_Error];
    }
}

- (void)flashAtmelDFU:(NSString *)mcu withFile:(NSString *)file {
    [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"erase", @"--force"]];
    NSString *result = [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"flash", @"--force", file]];
    if ([result containsString:@"Bootloader and code overlap."]) {
        [self.printer print:@"File is too large for device" withType:MessageType_Error];
    } else {
        [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"reset"]];
    }
}

- (void)resetAtmelDFU:(NSString *)mcu {
    [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"reset"]];
}

- (void)clearEEPROMAtmelDFU:(NSString *)mcu eraseFirst:(BOOL)erase {
    if (erase) {
        [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"erase", @"--force"]];
    }
    [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"flash", @"--force", @"--suppress-validation", @"--eeprom", @"reset.eep"]];
    if (erase) {
        [self.printer print:@"Please reflash device with firmware now" withType:MessageType_Bootloader];
    }
}

-(void)setHandednessAtmelDFU:(NSString *)mcu rightHand:(BOOL)rightHand eraseFirst:(BOOL)erase {
    NSString *file = [NSString stringWithFormat: @"reset_%@.eep", (rightHand ? @"right" : @"left")];
    if (erase) {
        [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"erase", @"--force"]];
    }
    [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"flash", @"--force", @"--suppress-validation", @"--eeprom", file]];
    if (erase) {
        [self.printer print:@"Please reflash device with firmware now" withType:MessageType_Bootloader];
    }
}

- (void)flashAtmelSAMBAwithFile:(NSString *)file {
    [self runProcess:@"mdloader" withArgs:@[@"-p", serialPort, @"-D", file, @"--restart"]];
}

- (void)resetAtmelSAMBA {
    [self runProcess:@"mdloader" withArgs:@[@"-p", serialPort, @"--restart"]];
}

- (void)flashAVRISP:(NSString *)mcu withFile:(NSString *)file {
    [self runProcess:@"avrdude" withArgs:@[@"-p", mcu, @"-c", @"avrisp", @"-U", [NSString stringWithFormat:@"flash:w:%@:i", file], @"-P", serialPort, @"-C", @"avrdude.conf"]];
}

- (void)flashBootloadHIDwithFile:(NSString *)file {
    [self runProcess:@"bootloadHID" withArgs:@[@"-r", file]];
}

- (void)resetBootloadHID {
    [self runProcess:@"bootloadHID" withArgs:@[@"-r"]];
}

- (void)flashCaterina:(NSString *)mcu withFile:(NSString *)file {
    [self runProcess:@"avrdude" withArgs:@[@"-p", mcu, @"-c", @"avr109", @"-U", [NSString stringWithFormat:@"flash:w:%@:i", file], @"-P", serialPort, @"-C", @"avrdude.conf"]];
}

- (void)clearEEPROMCaterina:(NSString *)mcu {
    [self runProcess:@"avrdude" withArgs:@[@"-p", mcu, @"-c", @"avr109", @"-U", @"eeprom:w:reset.eep:i", @"-P", serialPort, @"-C", @"avrdude.conf"]];
}

- (void)setHandednessCaterina:(NSString *)mcu rightHand:(BOOL)rightHand {
    NSString *file = [NSString stringWithFormat: @"reset_%@.eep", (rightHand ? @"right" : @"left")];
    [self runProcess:@"avrdude" withArgs:@[@"-p", mcu, @"-c", @"avr109", @"-U", [NSString stringWithFormat:@"eeprom:w:%@:i", file], @"-P", serialPort, @"-C", @"avrdude.conf"]];
}

- (void)flashHalfkay:(NSString *)mcu withFile:(NSString *)file {
    [self runProcess:@"teensy_loader_cli" withArgs:@[[@"-mmcu=" stringByAppendingString:mcu], file, @"-v"]];
}

- (void)resetHalfkay:(NSString *)mcu {
    [self runProcess:@"teensy_loader_cli" withArgs:@[[@"-mmcu=" stringByAppendingString:mcu], @"-bv"]];
}

- (void)flashKiibohdWithFile:(NSString *)file {
    if([[[file pathExtension] lowercaseString] isEqualToString:@"bin"]) {
        [self runProcess:@"dfu-util" withArgs:@[@"-D", file]];
    } else {
        [self.printer print:@"Only firmware files in .bin format can be flashed with dfu-util!" withType:MessageType_Error];
    }
}

- (void)flashLUFAMSwithFile:(NSString *)file {
    if (mountPoint != nil) {
        if ([[[file pathExtension] lowercaseString] isEqualToString:@"bin"]) {
            NSString *destFile = [NSString stringWithFormat:@"%@/FLASH.BIN", mountPoint];
            NSError *error;

            [self.printer print:[NSString stringWithFormat:@"Deleting %@...", destFile] withType:MessageType_Command];
            if (![[NSFileManager defaultManager] removeItemAtPath:destFile error:&error]) {
                [self.printer print:[NSString stringWithFormat:@"IO ERROR: %@", [error localizedDescription]] withType:MessageType_Error];
            }

            [self.printer print:[NSString stringWithFormat:@"Copying %@ to %@...", file, destFile] withType:MessageType_Command];
            if (![[NSFileManager defaultManager] copyItemAtPath:file toPath:destFile error:&error]) {
                [self.printer print:[NSString stringWithFormat:@"IO ERROR: %@", [error localizedDescription]] withType:MessageType_Error];
            }

            [self.printer print:@"Done, please eject drive now." withType:MessageType_Info];
        } else {
            [self.printer print:@"Only firmware files in .bin format can be flashed with this bootloader!" withType:MessageType_Error];
        }
    } else {
        [self.printer print:@"Could not find mount path for device!" withType:MessageType_Error];
    }
}

- (void)flashSTM32DFUWithFile:(NSString *)file {
    if([[[file pathExtension] lowercaseString] isEqualToString:@"bin"]) {
        [self runProcess:@"dfu-util" withArgs:@[@"-a", @"0", @"-d", @"0483:DF11", @"-s", @"0x8000000:leave", @"-D", file]];
    } else {
        [self.printer print:@"Only firmware files in .bin format can be flashed with dfu-util!" withType:MessageType_Error];
    }
}

- (void)flashSTM32DuinoWithFile:(NSString *)file {
    if([[[file pathExtension] lowercaseString] isEqualToString:@"bin"]) {
        [self runProcess:@"dfu-util" withArgs:@[@"-a", @"2", @"-d", @"1EAF:0003", @"-R", @"-D", file]];
    } else {
        [self.printer print:@"Only firmware files in .bin format can be flashed with dfu-util!" withType:MessageType_Error];
    }
}

- (void)flashUSBAsp:(NSString *)mcu withFile:(NSString *)file {
    [self runProcess:@"avrdude" withArgs:@[@"-p", mcu, @"-c", @"usbasp", @"-U", [NSString stringWithFormat:@"flash:w:%@:i", file], @"-C", @"avrdude.conf"]];
}

- (void)clearEEPROMUSBAsp:(NSString *)mcu {
    [self runProcess:@"avrdude" withArgs:@[@"-p", mcu, @"-c", @"usbasp", @"-U", @"eeprom:w:reset.eep:i", @"-C", @"avrdude.conf"]];
}

- (void)setHandednessUSBAsp:(NSString *)mcu rightHand:(BOOL)rightHand {
    NSString *file = [NSString stringWithFormat: @"reset_%@.eep", (rightHand ? @"right" : @"left")];
    [self runProcess:@"avrdude" withArgs:@[@"-p", mcu, @"-c", @"usbasp", @"-U", [NSString stringWithFormat:@"eeprom:w:%@:i", file], @"-C", @"avrdude.conf"]];
}

- (void)flashUSBTiny:(NSString *)mcu withFile:(NSString *)file {
    [self runProcess:@"avrdude" withArgs:@[@"-p", mcu, @"-c", @"usbtiny", @"-U", [NSString stringWithFormat:@"flash:w:%@:i", file], @"-C", @"avrdude.conf"]];
}
@end
